### PR TITLE
provide logic to check proxy settings and log a warning when a miscon…

### PR DESCRIPTION
**Pull request by bizob2828**
…figuration has possibly occurred

## Proposed Release Notes
 * Handle a proxy misconfiguration of collector and log an actionable warning message

## Links
Closes: #742 

## Details
 * To test agent with proxy server you can use this repo for the proxy https://github.com/bizob2828/node-proxy-server.
 *  Set `proxy_host` and `proxy_port` and watch the message occur
```
{"v":0,"level":40,"name":"newrelic","hostname":"C02F81D3MD6P","pid":33186,"time":"2021-05-03T14:39:06.028Z","msg":"Your proxy server appears to be configured to accept connections over http. When setting `proxy_host` and `proxy_port` New Relic attempts to connect over SSL(https). If your proxy is configured to accept connections over http, try setting `proxy` to a fully qualified URL(e.g http://proxy-host:8080).","component":"collector_api","errno":-100,"code":"EPROTO","syscall":"write","stack":"Error: write EPROTO 4429278720:error:1408F10B:SSL routines:ssl3_get_record:wrong version number:../deps/openssl/openssl/ssl/record/ssl3_record.c:332:\n\n    at WriteWrap.onWriteComplete [as oncomplete] (internal/stream_base_commons.js:94:16)\n    at WriteWrap.callbackTrampoline (internal/async_hooks.js:131:14)","message":"write EPROTO 4429278720:error:1408F10B:SSL routines:ssl3_get_record:wrong version number:../deps/openssl/openssl/ssl/record/ssl3_record.c:332:\n"}
```

### CODE REVIEW

File: test/unit/collector/api-connect.test.js [hunk 1]
```diff
@@ -591,7 +586,7 @@ tap.test('retries on receiving invalid license key (401)', (t) => {
     collectorApi = new CollectorApi(agent)
 
     const preconnectURL = helper.generateCollectorPath('preconnect')
-    failure = nock(URL).post(preconnectURL).times(5).reply(401, error)
+    failure = nock(URL).post(preconnectURL).times(5).reply(401)
```
Comment (bizob2828) [inline test/unit/collector/api-connect.test.js:589]: this error was superfluous

File: lib/collector/api.js [hunk 2]
```diff
@@ -117,6 +117,22 @@ CollectorAPI.prototype.connect = function connect(callback) {
...
+   * Checks if proxy is configured to connect via `proxy_host` and `proxy_port`
+   * and if error code is EPROTO or ECONNRESET. This is an indication their proxy
+   * server only accepts HTTP connections, and we should provide an actionable warning to
+   * fix the misconfiguration by setting `proxy` to a fully qualified URL
+   *
+   * @param {Error} error
+   * @returns {Boolean}
+   */
+  function isProxyMisconfigured(error) {
+    const config = api._agent.config || {}
```
Comment (bizob2828) [inline lib/collector/api.js:130]: this might be super defensive

File: test/unit/mocks/logger.js [hunk 3]
```diff
@@ -0,0 +1,10 @@
+'use strict'
+const sinon = require('sinon')
+
+module.exports = (sandbox = sinon) => ({
```
Comment (bizob2828) [inline test/unit/mocks/logger.js:4]: figured we could re-use this mock, updated the one place I was using it before

File: test/unit/collector/api-connect.test.js [hunk 4]
```diff
@@ -633,6 +628,94 @@ tap.test('retries on receiving invalid license key (401)', (t) => {
...
+
+    done()
+  })
+
+  t('should log warning when proxy is misconfigured', (t) => {
+    testConnect(t, () => {
+      t.deepEqual(loggerMock.warn.args, [
+        [
+          error,
+          'Your proxy server appears to be configured to accept connections over http. When setting `proxy_host` and `proxy_port` New Relic attempts to connect over SSL(https). If your proxy is configured to accept connections over http, try setting `proxy` to a fully qualified URL(e.g http://proxy-host:8080).'
```
Comment (carlo-808) [inline test/unit/collector/api-connect.test.js:694]: surprised linter didn't catch this. Possible to wrap text? or just put in `expectedErrorMsg` var?

File: test/unit/collector/api-connect.test.js [hunk 5]
```diff
@@ -633,6 +628,94 @@ tap.test('retries on receiving invalid license key (401)', (t) => {
...
+
+  t('should not log warning when proxy is configured properly but still get EPROTO', (t) => {
+    collectorApi._agent.config.proxy = 'http://test-proxy-server:8080'
+    testConnect(t, () => {
+      t.deepEqual(loggerMock.warn.args, [], 'Proxy misconfigured message not logged')
+      t.end()
+    })
+  })
+
+  function testConnect(t, cb) {
```
Comment (carlo-808) [inline test/unit/collector/api-connect.test.js:709]: nice!

File: test/unit/facts.test.js [hunk 6]
```diff
@@ -14,10 +14,7 @@ const expect = chai.expect
 const helper = require('../lib/agent_helper')
 const sinon = require('sinon')
 const proxyquire = require('proxyquire')
-const loggerMock = {
-  trace: sinon.stub(),
-  debug: sinon.stub()
-}
+const loggerMock = require('./mocks/logger')();
```
Comment (carlo-808) [inline test/unit/facts.test.js:17]: semicolon

File: test/unit/mocks/logger.js (hunk 3 above)
Comment (carlo-808) [reply test/unit/mocks/logger.js:4]: Nice add. Super helpful!

File: lib/collector/api.js (hunk 2 above)
Comment (carlo-808) [reply lib/collector/api.js:130]: Yeah, I think we always get at least the default config here.

**Review by carlo-808 — COMMENTED** [review]
Just a few minor comments but looks good overall. 👍 
There's a linting error that needs attention but I'll approve after that's resolved.

File: test/unit/facts.test.js (hunk 6 above)
Comment (bizob2828) [reply test/unit/facts.test.js:17]: def hitting this husky task next

File: test/unit/collector/api-connect.test.js (hunk 4 above)
Comment (bizob2828) [reply test/unit/collector/api-connect.test.js:694]: I think it's because we have a diff .eslintrc for tests

**Review by carlo-808 — APPROVED** [review]
👍

**Pull request merged**